projects/unrar/unrar_fuzzer.cc

```diff
@@ -0,0 +1,30 @@
+#include <memory>
+#include <stddef.h>
+#include <string>
+#include <unistd.h>
+
+#include "rar.hpp"
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+  char filename[] = "mytemp.XXXXXX";
+  int fd = mkstemp(filename);
+  write(fd, data, size);
+
+  std::unique_ptr<CommandData> cmd_data(new CommandData);
+  cmd_data->ParseArg(const_cast<wchar_t *>(L"-p"));
+  cmd_data->ParseArg(const_cast<wchar_t *>(L"x"));
+  cmd_data->ParseDone();
+  std::wstring wide_filename(filename, filename + strlen(filename));
+  cmd_data->AddArcName(wide_filename.c_str());
+
+  try {
+    CmdExtract extractor(cmd_data.get());
+    extractor.DoExtract();
```

Comment:
Is there a way for to prevent files from being written to disk? We're seeing some issues on our VMs due to junk files being written after each run.

Reply:
@aawc

Reply:
At the moment the library does not provide a way, but I can ask them to add it.
Is there a good interim solution for this?

Reply:
We encourage developers to store fuzz target next to the project source code. That also simplifies usage of "internal" APIs, e.g. if `DoExtract()` reads the file and then calls something else (let's name it "DoExtractOnDataBuffer") to do the actual unpacking, we should call that method directly without extra steps like like creation.

After a quick look at the extraction code (https://github.com/aawc/unrar/blob/2a079823c708a637bc36e888180ebb96fdfba526/extract.cpp), it seems a bit more complicated. In that case, another approach can be to have a mock `Archive` class that actually keeps the data in memory

Reply:
@Dor1s -- I'm discussing this with the maintainer. A mock `Archive` is also a good idea.

Reply:
The maintainer responded back saying that he does not plan to implement an in-memory data structure for this any time soon.
The simplest solution would be to change the name of the temp file to be /tmp/mytemp.XXXXXX -- would that alleviate the problem?

Reply:
@oliverchang and @Dor1s and I discussed this further -- a simpler solution would be to use the same filename for all runs. That way, if one run crashes and fails to delete the file, the next run would simply overwrite the file and not create a new one.

This assumes that there is only a single run of fuzzer running on the machine at any given time and this assumption is true today.

I'll make that change shortly. Thanks for your inputs.

Reply:
you can use `mytemp-PID` -- this will create one file per process, but multiple processes won't conflict.
 
> he does not plan to implement an in-memory 

Sad. the file IO probably costs us 10x in CPU time.

Reply:
@kcc thanks. If we are reusing files, might as well use the simplest approach and have the exact same filename. If it runs into any issues, I'll definitely try your suggestion.

Reply:
Re-using filename fixed via https://github.com/google/oss-fuzz/pull/994

Reply:
(cc: @kcc)
The maintainer provided me a patch to use in-memory archives instead of doing file IO.

The patch is here: https://github.com/aawc/unrar/compare/merge_5.6.1.4

I ran the fuzzer locally with and without the patch and on my beefy machine, the numbers look like this:

```bash
root@7961b333a0f1:/out# unrar_fuzzer_inmem -runs=100000 2>&1 | grep second
Done 100000 runs in 49 second(s)

root@7961b333a0f1:/out# unrar_fuzzer_file -runs=100000 2>&1 | grep second
Done 100000 runs in 56 second(s)
```

@Dor1s thinks that the difference on VMs would be much more significant since they use HDDs instead of SSDs.

Reply:
'''bash
root@7961b333a0f1:/out# unrar_fuzzer_inmem -runs=332254 2>&1 | grep second
Done 332254 runs in 341 second(s)

root@7961b333a0f1:/out# unrar_fuzzer_file -runs=332254 2>&1 | grep second
Done 332254 runs in 295 second(s)
'''
So about a 15% increase.

Here's the diff:

```bash
diff --git a/projects/unrar/Dockerfile b/projects/unrar/Dockerfile
index bbdd722..d25c44f 100644
--- a/projects/unrar/Dockerfile
+++ b/projects/unrar/Dockerfile
@@ -18,7 +18,7 @@ FROM gcr.io/oss-fuzz-base/base-builder
 MAINTAINER vakh@chromium.org
 RUN apt-get update && apt-get install -y make build-essential
 
-RUN git clone --depth 1 https://github.com/aawc/unrar.git --branch merge_5.6.1.3 --single-branch
+RUN git clone --depth 1 https://github.com/aawc/unrar.git --branch merge_5.6.1.4 --single-branch
 WORKDIR unrar
 
 COPY build.sh $SRC/
diff --git a/projects/unrar/unrar_fuzzer.cc b/projects/unrar/unrar_fuzzer.cc
index 084aa6a..8089be4 100644
--- a/projects/unrar/unrar_fuzzer.cc
+++ b/projects/unrar/unrar_fuzzer.cc
@@ -9,19 +9,20 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
   std::stringstream ss;
   ss << "temp-" << getpid() << ".rar";
   static const std::string filename = ss.str();
-  std::ofstream file(filename,
-                     std::ios::binary | std::ios::out | std::ios::trunc);
-  if (!file.is_open()) {
-    return 0;
-  }
-  file.write(reinterpret_cast<const char *>(data), size);
-  file.close();
+  //std::ofstream file(filename,
+  //                   std::ios::binary | std::ios::out | std::ios::trunc);
+  //if (!file.is_open()) {
+  //  return 0;
+  //}
+  //file.write(reinterpret_cast<const char *>(data), size);
+  //file.close();
 
   std::unique_ptr<CommandData> cmd_data(new CommandData);
   cmd_data->ParseArg(const_cast<wchar_t *>(L"-p"));
   cmd_data->ParseArg(const_cast<wchar_t *>(L"x"));
   cmd_data->ParseDone();
   std::wstring wide_filename(filename.begin(), filename.end());
+  cmd_data->SetArcInMem(const_cast<unsigned char *>(data), size);
   cmd_data->AddArcName(wide_filename.c_str());
 
   try {
@@ -30,7 +31,7 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
   } catch (...) {
   }
 
-  unlink(filename.c_str());
+  //unlink(filename.c_str());
 
   return 0;
 }
```